pkg/provider/provider_mux.go

```diff
@@ -0,0 +1,87 @@
+
+	"github.com/hewlettpackard/hpegl-provider-lib/pkg/registration"
+)
+
+// ProviderForMux is a function that returns a list of ProviderServer functions that can be used with the
+// Hashicorp mux library.  This function will be called from terraform-provider-hpegl which will be adapted
+// to support "legacy" provider code that use SDK v2.0 (i.e. metal, vmaas, caas on PCE) as well as newer provider
+// code that uses the new Hashicorp provider "framework".
+//
+// Note that we will need to add the ProviderSchemaEntry() functions for the newer providers.  This means that
+// registration.ServiceRegistration implementations for the newer providers that only contain ProviderSchemaEntry()
+// and no SupportedResource() or SupportedDataSources().
+func ProviderForMux(reg []registration.ServiceRegistration, pf ConfigureFunc) []func() tfprotov5.ProviderServer {
+	providerSchema := generateProviderSchema(reg)
+	providerServerList := make([]func() tfprotov5.ProviderServer, 0)
+	for _, service := range reg {
+		// Only create a provider if it has resources or data sources
+		if service.SupportedResources() != nil || service.SupportedDataSources() != nil {
+			providerServerList = append(providerServerList, generateProvider(service, pf, providerSchema))
+		}
+	}
+
+	return providerServerList
+}
+
+// generateProviderSchema generates the provider schema from the service registrations.  Note that this schema
+// needs to be added to each of the sub-providers.
+func generateProviderSchema(reg []registration.ServiceRegistration) map[string]*schema.Schema {
+	providerSchema := Schema()
+	for _, service := range reg {
+		if service.ProviderSchemaEntry() != nil {
+			// We panic if the service.Name() key is repeated in providerSchema
+			if _, ok := providerSchema[service.Name()]; ok {
+				panic(fmt.Sprintf("service name %s is repeated", service.Name()))
+			}
+			providerSchema[service.Name()] = convertToTypeSet(service.ProviderSchemaEntry())
+		}
+	}
+
+	return providerSchema
+}
+
+// generateProvider will generate a sub-provider for each service that can be used with the Hashicorp mux library.
+func generateProvider(
+	service registration.ServiceRegistration,
+	pf ConfigureFunc,
+	providerSchema map[string]*schema.Schema,
+) func() tfprotov5.ProviderServer {
+	dataSources := make(map[string]*schema.Resource)
+	resources := make(map[string]*schema.Resource)
+	// providerSchema is the Schema for the provider
+	for k, v := range service.SupportedDataSources() {
+		// We panic if the data-source name k is repeated in dataSources
+		if _, ok := dataSources[k]; ok {
+			panic(fmt.Sprintf("data-source name %s is repeated in service %s", k, service.Name()))
+		}
+		dataSources[k] = v
+	}
+	for k, v := range service.SupportedResources() {
+		// We panic if the resource name k is repeated in resources
+		if _, ok := resources[k]; ok {
+			panic(fmt.Sprintf("resource name %s is repeated in service %s", k, service.Name()))
+		}
+		resources[k] = v
+	}
```

Comment:
The the two range checks needed still? Previously dataSources/resources spanned multiple providers so they could clash, is a clash still possible?

Reply:
No and yes.  I'll push-up the change.